### ISSUE TITLE
Timed out trying to read data from the socket stream!

### ISSUE THREAD

**Issue by AlexandrKubit** [robinrodricks/FluentFTP#1193]
**FTP Server OS:**  Windows 

**FTP Server Type:** Windows Server IIS 

**Client Computer OS:** Windows 

**FluentFTP Version:** 44.0.1

**Framework:** .NET 7 

I ran into a problem. There is a timeout while downloading the file. This is only true for files with a long path (remote). At the same time, it somehow depends on the settings of the operating system, because. I don't see this issue on any other machine.

C# exception
```
/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Error while downloading the file from the server. See InnerException for more info.
   at FluentFTP.AsyncFtpClient.DownloadFileInternalAsync(String localPath, String remotePath, Stream outStream, Int64 restartPosition, IProgress`1 progress, CancellationToken token, FtpProgress metaProgress, Int64 knownFileSize, Boolean isAppend)
   at FluentFTP.AsyncFtpClient.DownloadFileToFileAsync(String localPath, String remotePath, FtpLocalExists existsMode, FtpVerify verifyOptions, IProgress`1 progress, CancellationToken token, FtpProgress metaProgress)
   at FluentFTP.AsyncFtpClient.DownloadFile(String localPath, String remotePath, FtpLocalExists existsMode, FtpVerify verifyOptions, IProgress`1 progress, CancellationToken token)
   at ACSUpdaterAPI.Services.FTPHandlingService.DownloadFtpFile(String localName, String remoteName, AsyncFtpClient client) in D:\Solforb\ACSUpdaterAPI\Services\FTPHandlingService.cs:line 315
Timed out trying to read data from the socket stream!
```

**Logs :**

```

Status:   Executing stashed command:
:
Command:  SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Status:   Waiting for response to: SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Status:   Closing/Disposing FtpSocketStream(control connection):
>         DownloadFile("C:\Solforb\TIGK\UpdateTemp\ASUCommon\Content\template\Шаблон загрузки продукций к договору.xlsx", "/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", Resume, None):
>         OpenRead("/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", Binary, 0, 0):
>         GetFileSize("/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", -1):
Warning:  Reconnect needed due to disconnected control connection (SslSessionLength: 0):
Status:   Command stashed: SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
>         Connect(True):
Warning:  Reconnect (Count: 20):
Status:   FluentFTP 44.0.1.0:
Status:   Connecting to IP #1= ***:21:
Status:   Waiting for a response:
Response: 220-FileZilla Server 1.6.6:
Response: 220 Please visit https://filezilla-project.org/ [1,668m]:
Status:   Detected FTP server: FileZilla:
Command:  USER ***:
Status:   Waiting for response to: USER ***:
Response: 331 Please, specify the password. [80ms]:
Command:  PASS ***:
Status:   Waiting for response to: PASS ***:
Response: 230 Login successful. [136ms]:
Command:  FEAT:
Status:   Waiting for response to: FEAT:
Response: 211-Features:
Response: MDTM
Response: REST STREAM
Response: SIZE
Response: MLST type*;size*;modify*;perm*;
Response: MLSD
Response: AUTH SSL
Response: AUTH TLS
Response: PROT
Response: PBSZ
Response: UTF8
Response: TVFS
Response: EPSV
Response: EPRT
Response: MFMT:
Response: 211 End [74ms]:
Status:   Text encoding: System.Text.UTF8Encoding+UTF8EncodingSealed:
Command:  OPTS UTF8 ON:
Status:   Waiting for response to: OPTS UTF8 ON:
Response: 202 UTF8 mode is always enabled. No need to send this command [48ms]:
Command:  SYST:
Status:   Waiting for response to: SYST:
Response: 215 UNIX emulated by FileZilla. [51ms]:
Status:   Listing parser set to: Machine:
Command:  TYPE I:
Status:   Waiting for response to: TYPE I:
Response: 200 Type set to I [57ms]:
>         SetWorkingDirectory("/"):
Command:  CWD /:
Status:   Waiting for response to: CWD /:
Response: 250 CWD command successful [58ms]:
Command:  PWD:
Status:   Waiting for response to: PWD:
Response: 257 "/" is current directory. [42ms]:
:
Status:   Executing stashed command:
:
Command:  SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Status:   Waiting for response to: SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Status:   Closing/Disposing FtpSocketStream(control connection):
>         DownloadFile("C:\Solforb\TIGK\UpdateTemp\ASUCommon\Content\template\Шаблон загрузки продукций к договору.xlsx", "/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", Resume, None):
>         OpenRead("/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", Binary, 0, 0):
>         GetFileSize("/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", -1):
Warning:  Reconnect needed due to disconnected control connection (SslSessionLength: 0):
Status:   Command stashed: SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
>         Connect(True):
Warning:  Reconnect (Count: 21):
Status:   FluentFTP 44.0.1.0:
Status:   Connecting to IP #1= ***:21:
Status:   Waiting for a response:
Response: 220-FileZilla Server 1.6.6:
Response: 220 Please visit https://filezilla-project.org/ [1,668m]:
Status:   Detected FTP server: FileZilla:
Command:  USER ***:
Status:   Waiting for response to: USER ***:
Response: 331 Please, specify the password. [47ms]:
Command:  PASS ***:
Status:   Waiting for response to: PASS ***:
Response: 230 Login successful. [114ms]:
Command:  FEAT:
Status:   Waiting for response to: FEAT:
Response: 211-Features:
Response: MDTM
Response: REST STREAM
Response: SIZE
Response: MLST type*;size*;modify*;perm*;
Response: MLSD
Response: AUTH SSL
Response: AUTH TLS
Response: PROT
Response: PBSZ
Response: UTF8
Response: TVFS
Response: EPSV
Response: EPRT
Response: MFMT:
Response: 211 End [55ms]:
Status:   Text encoding: System.Text.UTF8Encoding+UTF8EncodingSealed:
Command:  OPTS UTF8 ON:
Status:   Waiting for response to: OPTS UTF8 ON:
Response: 202 UTF8 mode is always enabled. No need to send this command [51ms]:
Command:  SYST:
Status:   Waiting for response to: SYST:
Response: 215 UNIX emulated by FileZilla. [40ms]:
Status:   Listing parser set to: Machine:
Command:  TYPE I:
Status:   Waiting for response to: TYPE I:
Response: 200 Type set to I [46ms]:
>         SetWorkingDirectory("/"):
Command:  CWD /:
Status:   Waiting for response to: CWD /:
Response: 250 CWD command successful [42ms]:
Command:  PWD:
Status:   Waiting for response to: PWD:
Response: 257 "/" is current directory. [46ms]:


```

FileZilla Server
```
"Date","Info","Type","Message"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1","Response","220-FileZilla Server 1.6.6"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1","Response","220 Please visit https://filezilla-project.org/"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1","Command","USER FTPConnection"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1","Response","331 Please, specify the password."
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1","Command","PASS ****"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","230 Login successful."
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Command","FEAT"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","211-Features:"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","211 End"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Command","OPTS UTF8 ON"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","202 UTF8 mode is always enabled. No need to send this command"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Command","SYST"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","215 UNIX emulated by FileZilla."
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Command","TYPE I"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","200 Type set to I"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Command","CWD /"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","250 CWD command successful"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Command","PWD"
"08-02-2023 15:35:28","FTP Session 154 10.0.0.1 FTPConnection","Response","257 ""/"" is current directory."
"08-02-2023 15:35:29","FTP Session 154 10.0.0.1 FTPConnection","Command","SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx"
"08-02-2023 15:35:29","FTP Session 154 10.0.0.1 FTPConnection","Response","213 11353"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1","Response","220-FileZilla Server 1.6.6"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1","Response","220 Please visit https://filezilla-project.org/"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1","Command","USER FTPConnection"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1","Response","331 Please, specify the password."
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1","Command","PASS ****"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","230 Login successful."
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","FEAT"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","211-Features:"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","211 End"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","OPTS UTF8 ON"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","202 UTF8 mode is always enabled. No need to send this command"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","SYST"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","215 UNIX emulated by FileZilla."
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","TYPE I"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","200 Type set to I"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","CWD /"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","250 CWD command successful"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","PWD"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","257 ""/"" is current directory."
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Command","SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx"
"08-02-2023 15:37:09","FTP Session 155 10.0.0.1 FTPConnection","Response","213 11353"
```

C# code
```
bool downloaded = false;

Console.WriteLine(remoteName + "  started");

if(await client.DownloadFile(localName, remoteName) == FtpStatus.Success)
{
    asyncFtpClients.Add(client);
    downloaded = true;
    Console.WriteLine(remoteName + "  success");
}
else
{
    LogFailed(remoteName);
}
```

**Comment by FanDjango** [comment]
You are getting very numerous automatic reconnects due to an unexpectedly dropped control connection.
```
Command:  SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Status:   Waiting for response to: SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
Status:   Closing/Disposing FtpSocketStream(control connection):
>         DownloadFile("C:\Solforb\TIGK\UpdateTemp\ASUCommon\Content\template\Шаблон загрузки продукций к договору.xlsx", "/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", Resume, None):
>         OpenRead("/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", Binary, 0, 0):
>         GetFileSize("/ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx", -1):
Warning:  Reconnect needed due to disconnected control connection (SslSessionLength: 0):
Status:   Command stashed: SIZE /ASUCommon/Content/template/Шаблон загрузки продукций к договору.xlsx:
>         Connect(True):
Warning:  Reconnect (Count: 20):
Status:   FluentFTP 44.0.1.0:
Status:   Connecting to IP #1= ***:21:
Status:   Waiting for a response:
Response: 220-FileZilla Server 1.6.6:
```

You should have a look at the **first** occurence of that (Count: 1) and see how the problem develops from there. You are connecting to a FileZilla server, perhaps in the log of the server you can see the reason he drops the connection?

**Comment by FanDjango** [comment]
>  the method is recursive, if it catches an exception, it calls itself.

Ok, that explains the reconnects.

> It feels like the client can't get the file size.

I agree, but why?

1. The file name is not overly long
2. FileZilla seems to be sending a ""213 11353" response
3. It does not seem to arrive at FluenFTP and 90 seconds later, there is a reconnect. I don't see a FluentFTP timeout exception?

**Comment by alexandrkubit** [comment]
I also want to note that loading from another machine occurs without problems. That is, my application runs on another machine and connects to the same FTP server and this file is downloaded. I do not know where to dig.

client config: 
```
                var client = new AsyncFtpClient(ftpAddress, credentials, 21);
                client.LegacyLogger = OnFTPLogEvent;
                client.Config.EncryptionMode = FtpEncryptionMode.None;

                await client.Connect();
                client.Config.ConnectTimeout = 100000;
                client.Config.ReadTimeout = 100000;
                client.Config.DataConnectionConnectTimeout = 100000;
                client.Config.DataConnectionReadTimeout = 100000;
```

C# exception :  Timed out trying to read data from the socket stream!

**Comment by FanDjango** [comment]
Yeah, I see.
Could it be a content aware firewall just "killing" the connection? Perhaps the "213... " response in the FileZilla log is logged, but never reaches the other end. The firewall might already be closing the connection as a reaction to the file name.

**Comment by alexandrkubit** [comment]
I also tried to deploy a server on IIS, the same story. The firewall is completely disabled on all machines. And I tried different names, the essence is only in the total length of their path, for example, the file /ASUCommon/Content/1234567890123/Product download template to contract.xlsx is loaded, but /ASUCommon/Content/1234567890123/Product download template to contract.xlsx is not. I think that this is some kind of setting of the operating system, but I don’t know which one.
